emulator/src/extensions/gamepad/ext_gamepad.c

```diff
@@ -0,0 +1,628 @@
+}
+
+bool gamepadInitCb(emuArgs_t* args)
+{
+    if (gamepadConnect(&joystickExt))
+    {
+        ESP_LOGI("GamepadExt", "Connected to joystick with %d axes and %d buttons\n", joystickExt.numAxes,
+                 joystickExt.numButtons);
+        return true;
+    }
+
+    return false;
+}
+
+void gamepadDeinitCb(void)
+{
+    gamepadDisconnect(&joystickExt);
+}
+
+void gamepadPreFrameCb(uint64_t frame)
+{
+    emuJoystickEvent_t event;
+    while (gamepadReadEvent(&joystickExt, &event))
+    {
+        switch (event.type)
+        {
+            case BUTTON:
+            {
+                if (event.button < 32 && joystickButtonMap[event.button])
+                {
+                    emulatorInjectButton(joystickButtonMap[event.button], event.value != 0);
+                }
+                break;
+            }
+
+            case AXIS:
+            {
+                switch (event.axis)
+                {
+                    case TOUCH_AXIS_X:
+                    case TOUCH_AXIS_Y:
+                    {
+                        // Joystick data comes in as a value from (-2**15) to (2**15-1)
+                        // We need to take that down to (-128) to (127) (divide by 8)
+                        // Then need to take the atan2 of it
+                        double x = (double)joystickExt.axisData[TOUCH_AXIS_X];
+                        double y = (double)joystickExt.axisData[TOUCH_AXIS_Y];
+
+                        if (x == 0 && y == 0)
+                        {
+                            emulatorSetTouchJoystick(0, 0, 0);
+                        }
+                        else
+                        {
+                            const double maxRadius = 1024;
+
+                            double angle        = atan2(-y, x);
+                            double angleDegrees = (angle * 180) / M_PI;
+                            int radius          = CLAMP(sqrt(x * x + y * y) * maxRadius / 32768, 0, maxRadius);
+
+                            int angleDegreesRounded = (int)(angleDegrees + 360) % 360;
+                            emulatorSetTouchJoystick(CLAMP(angleDegreesRounded, 0, 359), radius, 1024);
+                        }
+                        break;
+                    }
+
+                    case ACCEL_AXIS_X:
+                    case ACCEL_AXIS_Y:
+                    case ACCEL_AXIS_Z:
+                    {
+                        int16_t accelX = joystickExt.axisData[ACCEL_AXIS_X] / 128;
+                        int16_t accelY = joystickExt.axisData[ACCEL_AXIS_Y] / 128;
+                        int16_t accelZ = joystickExt.axisData[ACCEL_AXIS_Z] / 128;
+
+                        emulatorSetAccelerometer(accelX, accelY, accelZ);
+                        break;
+                    }
+
+                    case DPAD_AXIS_X:
+                    case DPAD_AXIS_Y:
+                    {
+                        buttonBit_t curState = 0;
+                        if (joystickExt.axisData[DPAD_AXIS_X] < 0)
```

Comment:
So on the actual Swadge the D-Pad is four discrete buttons, but here's it's a proper D-Pad, so the hardware can handle simultaneous up/down presses while the joystick cannot. I don't think it matters, just a point of interest.

Reply:
Ah, yeah that's true, probably not something we'd realistically need to support anyway. This is also just how the swadge maps the D-pad in gamepad mode, and I guess it's pretty standard since a switch controller also maps its D-pad the same way.